keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md

```diff
@@ -0,0 +1,691 @@
+- [Alternatives](#alternatives)
+  - [How the configuration is stored](#how-the-configuration-is-stored)
+  - [How the configuration is updated](#how-the-configuration-is-updated)
+  - [Support for future registry topologies](#support-for-future-registry-topologies)
+  - [Support for other local image loaders](#support-for-other-local-image-loaders)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+Local clusters like Kind, K3d, Minikube, and Microk8s let users iterate on Kubernetes
+quickly in a hermetic environment. To avoid network round-trip latency, these
+clusters can be configured to pull from a local, insecure registry. 
```

Comment:
OpenShift provides a local secured registry by default. It's secured by a self-signed CA, but all internal components know that the registry uses this CA, and this CA may be exported for external components.

Is there any reason not to add `insecure` and `caBundle` to the proposed config?

Reply:
@dmage Oooh that's a deep question! I don't think there's any active harm in putting it there. But what would you expect to happen differently if we did? i.e., what should we tell the consumer of the ConfigMap to do with this info? 

I can think of a few different possibilities!

1) The cluster configuration script has already set up your local docker config / registries.conf with the caBundle so that pushing to the registry will work out-of-the-box. So this is just an FYI to the IDE / dev env reading this ConfigMap. They don't need to take any action.

2) When the IDE / dev env reads the ConfigMap, it's responsible for updating the docker config / registries.conf itself with the caBundle (if it can).

3) When the IDE / dev env reads the ConfigMap, it's responsible for checking that the docker config / registries.conf has this information, and refusing to push to that registry if it's not and/or giving the user instructions on how to set it up.

I think I was assuming that we live in world (1)  - that the tooling setting up the local cluster has enough information to set up the configs itself, and doesn't need to push the burden downstream to other tools (or even worse, pushing it onto the human)

Reply:
friendly ping in case you missed this comment

the more i think about it, we're probably better off if we omit the fields, rather than add the fields and all consumers ignore them.

maybe a better solution is to add some text in the spec like "if pushing to the registry fails, the consumer should include the `help` link in the error message to the user". Then cluster providers can use the `help` link to show users how to configure the registry caBundle

Reply:
> Is there any reason not to add insecure and caBundle to the proposed config?

some questions from me here:
- when a CA certificate is exposed in a public space (e.g. `kube-public` NS) this can be used for clients to verify the server they are trying to reach is the one they want. this gives clients the chance to trust this server. but given anyone can get the CA, the server cannot trust all clients. is that the intention here?
- what is the expected / suggested behavior if `insecure` (assuming it's a bool) is `false` but `caBundle` is empty?
- would `caBundle` be base64 encoded?